server/src/main/java/io/crate/planner/operators/Insert.java

```diff
@@ -90,12 +92,24 @@ public ExecutionPlan build(DependencyCarrier executor,
         var boundIndexWriterProjection = writeToTable
             .bind(x -> SubQueryAndParamBinder.convert(x, params, subQueryResults));
 
-        if (sourcePlan.resultDescription().hasRemainingLimitOrOffset()) {
+        ResultDescription resultDescription = sourcePlan.resultDescription();
+        if (resultDescription.hasRemainingLimitOrOffset()) {
             ExecutionPlan localMerge = Merge.ensureOnHandler(sourcePlan, plannerContext);
             localMerge.addProjection(boundIndexWriterProjection);
             return localMerge;
         } else {
-            sourcePlan.addProjection(boundIndexWriterProjection);
+            // this also clears the positional order by if it is present.
+            // Otherwise the following `ensureOnHandler` would try to do a sorted-merge
+            // which doesn't work after boundIndexWriterProjection becuase it returns only
+            // rowcounts in the first column
+            //
+            // Ignoring the ordering is safe as long as there is no limit (which is implied, given `hasRemainingLimitOrOffset` was false)
+            sourcePlan.addProjection(
```

Comment:
Can't we have limit even if `hasRemainingLimitOrOffset` was false?

`return limit() != LimitAndOffset.NO_LIMIT || offset() != 0`  can return false if we have limit but have `offset = 0.`

Also, semantics of using
```
LimitAndOffset.NO_LIMIT,
LimitAndOffset.NO_OFFSET
```
look like that we have no limit AND no offset but we can end up in this branch if at least one is there.

Reply:
> Can't we have limit even if hasRemainingLimitOrOffset was false?

It's an `OR`. If this is branch is hit there is no limit and no offset. 

```
limit() != LimitAndOffset.NO_LIMIT || offset() != 0

limit=10 offset=2
true || true -> true

limit=10 offset=0
true || false -> true

limit=-1 offset=1
false || true -> true

limit=-1 offset=0
false || false -> false
```

Reply:
right, I messed up OR-s with negations